primo/utils/config_utils.py

```diff
@@ -606,67 +616,525 @@ def __init__(
+            ][self.wd._col_names.cluster].item()
+            selections_dict.setdefault(cluster, []).append(well_index)
+            selections_list.append(well_index)
+        return selections_dict, selections_list
+
+    def return_value(self):
+        """A wrapper for returning selections of all widgets"""
+        well_removed_dict, well_removed_list = self._return_well_index_cluster(
+            self.remove_widget.return_selections()
+        )
+        remove_widget_return = WidgetReturn(
+            self.remove_widget.cluster_widget.return_selections(),
+            well_removed_dict,
+        )
+
+        existing_clusters, new_clusters = self.add_widget.return_selections()
+        well_add_dict, _ = self._return_well_index_cluster(existing_clusters)
+        add_widget_return = AddWidgetReturn(well_add_dict, new_clusters)
+
+        cluster_lock_list = self.lock_widget.cluster_widget.return_selections()
+        well_lock_dict, _ = self._return_well_index_cluster(
+            self.lock_widget.return_selections()
+        )
+        for cluster in cluster_lock_list:
+            well_lock_dict[cluster] = [
+                well
+                for well in self.opt_campaign[cluster]
+                if well not in well_removed_list
+            ]
+
+        lock_widget_return = WidgetReturn(cluster_lock_list, well_lock_dict)
+
+        return OverrideSelections(
+            remove_widget_return=remove_widget_return,
+            add_widget_return=add_widget_return,
+            lock_widget_return=lock_widget_return,
+        )
+
+
+@dataclass
+class WidgetReturn:
+    """
+    Class for storing the widget return
+
+    Parameters
+    ----------
+    cluster : List
+        List of projects that are removed or locked
+
+    well : Dict
+        Dictionary of list of wells that are removed or locked in a cluster
+    """
+
+    cluster: List[int]
+    well: Dict[int, List[int]]
+
+    def __str__(self):
+        return f"WidgetReturn(project={self.cluster}, " f"well_dict={self.well})"
+
+
+@dataclass
+class AddWidgetReturn:
+    """
+    Class for storing the add widget return
+
+    Parameters
+    ----------
+    existing_cluster : Dict
+        Dictionary of list of wells being added and their original cluster;
+        key=> cluster, value=> well list
+
+    new_cluster : Dict
+        Dictionary of list of wells being added and their new cluster;
+        key=> cluster, value=> well list
+    """
+
+    existing_cluster: Dict[int, List[int]]
+    new_cluster: Dict[int, List[int]]
+
+    def __str__(self):
+        return (
+            f"AddWidgetReturn(existing_cluster_dict={self.existing_cluster}, "
+            f"new_cluster_dict={self.new_cluster})"
+        )
+
+
+@dataclass
+class OverrideSelections:
+    """
+    Class for storing the return from the remove_widget, add_widget, and lock_widget
+
+    Parameters
+    ----------
+    remove_widget_return : WidgetReturn
+        Object returned by the remove_widget
```

Comment:
This all looks good. One possible suggestion (up for debate) is to change the names of the WidgetReturn and the AddWidgetReturn classes. There is nothing tying them to a widget and I use them in the API where there is no widget. So a more general name could be "OverrideRemoveLockInformation" and "OverrideAddInformation" or something along those lines.

Reply:
"OverrideSelections" is spot on.

Reply:
Thank you for the comments, the class names have been changed to `OverrideRemoveLockInfo` and `OverrideAddInfo`.